src/store_client.cc

```diff
@@ -717,15 +723,14 @@ storeUnregister(store_client * sc, StoreEntry * e, void *data)
 
 #endif
 
+    // We must lock to safely dereference e below, after deleting sc and after
+    // calling CheckQuickAbortIsReasonable().
+    e->lock("storeUnregister");
+
     // XXX: We might be inside sc store_client method somewhere up the call
     // stack. TODO: Convert store_client to AsyncJob to make destruction async.
     delete sc;
 
-    assert(e->locked());
```

Comment:
This problematic assertion was removed instead of moved earlier/higher because we now know that the entry is locked at least up until the `delete sc` call above. Moving this assertion before that call would be pointless/misleading. After that call, we cannot really check anything because `e` may be gone -- that is why we must lock it ourselves before making that delete call. The official code simply assumed that deleting `sc` cannot unlock an entry, and now that assumption is wrong.

We could assert that the lock count is at least 2, but counting locks like that is conceptually wrong, especially since this particular code does not really care whether `e` remains alive after we are done with it (and that is the only thing this problematic assert was enforcing).